types/hapi__joi/index.d.ts

```diff
@@ -608,6 +609,13 @@ declare namespace Joi {
     }
 
     interface ValidationError extends Error {
+        // Per the DefinitelyTyped README, classes should be preferred over new-able interfaces
+        // HOWEVER, changing this interface to a class breaks this definition file
+        // Discussion around this can be see in this pull request: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43970
+        // I really want a better solution, but this is better than nothing:
+        // tslint:disable-next-line:no-misused-new
+        new(): ValidationError;
```

Comment:
Unfortunately this is incorrect, this would would allow you to use the `new` operator on an instance rather than on the class itself (which still doesn't exist).

I can see how you came up with this tho, `Root` incorrectly declares `ValidationError: ValidationError;`.
joi does not have a member `ValidationError` that is a validation error, it's the *class* `ValidationError`.

See:
![image](https://user-images.githubusercontent.com/3259089/79611553-b571fd00-80af-11ea-8443-ff1bf7b28fa3.png)

Reply:
I get the potential error of allowing the `new` operator to be used on an instance, but I don't think it's accurate that ***nothing*** should be visible when typing `Joi.ValidationError` into a TypeScript file, because there ***is*** a class with that name. As you just demonstrated in Node, `Joi.ValidationError` exists and, since it's a class, the following code is perfectly valid:

```
const Joi = require("@hapi/joi");
const myValidationError = new Joi.ValidationError();
```

After some testing I figured out I can emulate this behavior with the following change to Root:

```
ValidationError: new () => ValidationError;
```

It's not ideal (extremely ugly, actually) but it gets the types to behave the way you would expect.

Along with your other comments, I've been working on some additional local changes to get everything in sync with v17.1 (the version that added the `isError` method:

 - Replaced `wrapArrays` with the new `wrap` object containing `array` and `label` members
 - Updated the `default` method parameters to reflect the changes in v17
 - Added the `regex` method
 - Updated tests accordingly

I'll make a new commit shortly after running lint and test and everything else

Reply:
Unfortunately this is a shortcoming of TS not being able to map out all the existing (and sometimes crazy) export patterns people in JS, even the node typings have exceptions like this 😛